### ISSUE TITLE
Add OCI Registry as a State Storage  

### ISSUE THREAD

**Issue by shanduur** [opentofu/opentofu#1230]
### OpenTofu Version

```shell
OpenTofu v1.6.1
on darwin_arm64
```

### Use Cases

The end goal I aim to achieve with requesting OCI Registry support as a remote state backend storage for OpenTofu is to enhance the security, scalability, and flexibility of managing infrastructure as code (IaC) deployments. By utilizing OCI Registry, which provides a secure and scalable container image registry, as a backend for storing OpenTofu state files, several use cases can be addressed:

1. **Security**: OCI Registry offers robust security features such as access control, image signing, and vulnerability scanning. By leveraging OCI Registry for OpenTofu state storage, sensitive infrastructure configuration information is kept within a secure container registry environment, reducing the risk of unauthorized access or tampering.

2. **Scalability**: OCI Registry is designed to handle large-scale container deployments, making it suitable for managing OpenTofu state files for complex infrastructure environments. As infrastructure deployments grow in size and complexity, having a scalable backend like OCI Registry ensures efficient storage and retrieval of OpenTofu state information.

3. **Versioning and Rollback**: OCI Registry supports versioning of container images, allowing for easy rollback to previous versions if necessary. Integrating OpenTofu with OCI Registry as a backend enables versioning of OpenTofu state files, facilitating reliable infrastructure deployment rollbacks and change tracking.

4. **Integration with Containerized Workflows**: For organizations adopting containerized workflows, utilizing OCI Registry for OpenTofu state storage aligns well with existing container-based infrastructure management practices. It enables seamless integration between OpenTofu-managed infrastructure and containerized applications, streamlining the overall deployment process.

5. **Multi-Cloud Deployments**: OCI Registry is cloud-agnostic and can be deployed across various cloud providers or on-premises environments. This flexibility makes it suitable for organizations operating in multi-cloud or hybrid cloud environments. By using OCI Registry as a backend for OpenTofu state storage, organizations can maintain a consistent approach to infrastructure management across different cloud platforms.

Overall, integrating OCI Registry support as a remote state backend for OpenTofu addresses key requirements around security, scalability, versioning, and multi-cloud deployment, empowering organizations to efficiently manage their infrastructure deployments with OpenTofu in container-centric environments.

### Attempted Solutions

N/A

### Proposal

Add new implementation of remote-state backend.

**Comment by apparentlymart** [comment]
This issue is essentially blocked on a decision for https://github.com/opentofu/opentofu/issues/382 right now, since we're not intending to add any new backend implementations until we have a more specific plan for how to maintain them sustainably. (Making the plugins isn't necessarily the way we will do that, but that's the main anchor issue we have for this overall problem.)

However, I think regardless of whether this would end up being a new built-in backend or a plugin-based backend, there are some design questions we'd need to address that relate to what we can expect from typical OCI Registry implementations. We've recently been working on https://github.com/opentofu/opentofu/issues/2540 which has created some foundations for OpenTofu to interact with OCI Registries, but we'd still need to decide exactly how OpenTofu would implement _state storage_.

I have at least two concrete questions that I think we'd need to address in designing this:

1. Can we assume that all OCI Distribution implementations offer read-after-write consistency for tags?

    In other words: is it guaranteed that a request to update a tag immediately followed by a request to read that same tag will return what was just written, even if the read request came from a different location than the write request? Some large-scale systems use an eventual consistency model where a read shortly after a write can potentially return a stale result, if e.g. the reader is relying on a local cache that wasn't yet invalidated.

2. How would we represent an advisory lock on a particular state snapshot?

    OpenTofu currently expects state storage backends to be able to support a mutual-exclusion model by allowing one process to acquire a lock and other concurrent processes to notice that the lock is held and not begin any operations that could require writing to the store. OCI Distribution only specifies _storage_, and doesn't offer a locking mechanism.

Overall I'm concerned that although systems like Amazon S3 are blob stores and OCI Distribution is also a blob store, OCI Distribution was designed for different use-cases and usage patterns: it's primarily focused on write-once-read-many distribution of never-changing or infrequently-changing software packages, rather than as a general-purpose datastore. There are also many different implementations of this protocol designed under different scale assumptions, and so I think we need to assume that anything that isn't directly required by the specification is likely to vary between implementations.

If anyone has thoughts on the above topics I think it's reasonable to discuss them here even though we're not yet sure whether a final implementation would be built in or plugin-based, since even a plugin-based implementation would need to answer similar questions and so could learn from whatever is the outcome of discussion here.

**Comment by stevehipwell** [comment]
@apparentlymart I think your questions and the blocker are valid if we're considering OCI registry storage as a general purpose alternative to object stores. But what if we're looking at it (at least initially) as a constrained solution that's more of an enabling pattern? An OCI registry backend would allow anyone with a GitHub (or any other SCM with built in OCI registry support) to start using OpenTofu as long as they're willing to live with some constraints (e.g. single actor). This would make OpenTofu with a remote backend instantly an option for many more people that don't have the money or time to configure an additional external system.

As an example if I create a GitHub workflow with concurrency control that is the only system with access to my OCI registry for state, then I don't need OpenTofu to manage locking and the consistency model would match that required for OCI image operations. This is a very low friction way to start using OpenTofu.

FYI based on quickly grepping the spec I think that an OCI registry that implements the referrers API would support locking.

**Comment by apparentlymart** [comment]
Thanks for that extra context, @stevehipwell.

If we can convince ourselves that it's okay to add a new backend that doesn't support locking at all (which, as far as I know, hasn't happened since long before OpenTofu existed) then I agree that could be a way to dismiss my question 2 by answering "we wouldn't"!

I don't think that necessarily addresses question 1, though. OpenTofu assumes that if one process A writes a state snapshot and then another process B (possibly running on a different computer) reads the latest state snapshot then B will definitely receive observe the state snapshot written by A. That constraint is not necessarily met by an "eventually consistent" system where e.g. responses are produced from a cache that's invalidated asynchronously from incoming write requests.

If we're willing to accept that this is really just a "GitHub Container Registry" backend and not a general "OCI Distribution protocol" backend then we do at least only need to convince ourselves that _GitHub_ guarantees read-after-write consistency for tags in its implementation, which is more practical. I don't know if GitHub _does_ make that guarantee, but at least it's a question of finite scope that we could try to answer. 😀

**Comment by stevehipwell** [comment]
> If we can convince ourselves that it's okay to add a new backend that doesn't support locking at all (which, as far as I know, hasn't happened since long before OpenTofu existed) then I agree that could be a way to dismiss my question 2 by answering "we wouldn't"!

@apparentlymart when you put it that way it sounds crazy*! But what if instead of comparing this to existing remote backends we compare it to local state? OCI could be a readily available distributed backend replacement for local state; replacing the hacks that store the state in Git, or store no state and import everything on plan. I do think we could also answer points 1 & 2 but that they're not important in the alternate context. Essentially OCI is a "good enough" solution for a specific use case but if you have access to actual infra you have access to object storage so use that!

> I don't think that necessarily addresses question 1, though. OpenTofu assumes that if one process A writes a state snapshot and then another process B (possibly running on a different computer) reads the latest state snapshot then B will definitely receive observe the state snapshot written by A. That constraint is not necessarily met by an "eventually consistent" system where e.g. responses are produced from a cache that's invalidated asynchronously from incoming write requests.
>
> If we're willing to accept that this is really just a "GitHub Container Registry" backend and not a general "OCI Distribution protocol" backend then we do at least only need to convince ourselves that GitHub guarantees read-after-write consistency for tags in its implementation, which is more practical. I don't know if GitHub does make that guarantee, but at least it's a question of finite scope that we could try to answer. 😀

If the referrers API I mentioned in my previous reply is implemented then wouldn't this solve that problem as to check the lock the system needs to be consistent?

_\* How many people were/are using S3 without locking though?_

**Comment by stevehipwell** [comment]
> If you already have a specific idea in mind for how the referrers API would be useful in this context then I'd love to hear more about it! Thanks.

@apparentlymart sorry I'm coming at this problem from the other end (a solution for declaratively managing GitHub orgs with Terraform/OpenTofu) so this isn't currently a yak I feel that I can shave. I'll probably test out my hypothesis using ORAS to store local state in an OCI registry while I wait to see how this goes.

**Comment by apparentlymart** [comment]
Sure thing... I only intended that as "if you already know an answer", and was not intending it as a request that you go do a bunch of original research. Anything you _do_ learn on your travels that's related to this would be appreciated, though!

**Comment by vmvarela** [comment]
Hi @apparentlymart, I've implemented a preliminary version of this backend (https://github.com/opentofu/opentofu/pull/3600) and I'm successfully using it with GHCR with a locking system based on lock/unlock tags per workspace. As @stevehipwell mentioned, it has allowed me to work with OpenTofu in GitHub Actions without requiring external infrastructure.

**Comment by apparentlymart** [comment]
Thanks for working on this, @vmvarela.

Note that this issue is still on hold awaiting a decision on https://github.com/opentofu/opentofu/issues/382. That issue is in turn waiting for more information on a new plugin protocol feature that the Terraform team appears to be working on, based on some preliminary items that have been added to the protocol description.

The most ideal outcome from our perspective is that someone who is interested in developing a state storage implementation for OCI Distribution can do it just once as a plugin that works for both OpenTofu and Terraform, instead of needlessly implementing essentially the same thing twice, but we need to understand the details of this new protocol feature before we can evaluate whether it's suitable for OpenTofu's purposes. (For example, OpenTofu supports client-side state encryption, which imposes a few additional requirements that Terraform doesn't have.)

We intend to return to this family of issues related to state storage only once the new protocol features are finalized.

**Comment by vmvarela** [comment]
Thanks for the detailed explanation @apparentlymart.
It makes perfect sense to avoid duplicate effort and wait for clarity on the new protocol features. I'll keep this PR on hold and monitor #382. Once the requirements are finalized, I'm happy to revisit this implementation or adapt it as needed.

**Comment by vmvarela** [comment]
Implemented support for compression, workspaces, and versioning (#3600). Awaiting final decision.
